flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -411,26 +416,30 @@ protected static <KEY, Context extends JobAutoScalerContext<KEY>> int scale(
 
         var numKeyGroupsOrPartitions =
                 numSourcePartitions <= 0 ? maxParallelism : numSourcePartitions;
-        var upperBoundForAlignment =
-                Math.min(
-                        // Optimize the case where newParallelism <= maxParallelism / 2
-                        newParallelism > numKeyGroupsOrPartitions / 2
-                                ? numKeyGroupsOrPartitions
-                                : numKeyGroupsOrPartitions / 2,
-                        upperBound);
+
+        KeyGroupOrPartitionsAdjustMode mode =
+                context.getConfiguration().get(SCALING_KEY_GROUP_PARTITIONS_ADJUST_MODE);
+
+        var upperBoundForAlignment = Math.min(numKeyGroupsOrPartitions, upperBound);
```

Comment:
Why the `upperBoundForAlignment` logic is updated? Would you mind sharing one case?

Reply:
The optimization to use half of the keyspace when the parallelism is less/equal to the keyspace doesn't work for the new parallelism adjustment mode. It was anyways just a shortcut, to avoid checking all divisors until the maximum. No harm removing it.

Reply:
> Why the `upperBoundForAlignment` logic is updated? Would you mind sharing one case?

When I ran the test case, i found  this logic is wrong when keygroups or partitionnum are not even numbers.

Reply:
It's seems it's wrong logic when numKeyGroupsOrPartitions is odd number. 

Thanks for the clarification! This change makes sense to me.